src/typechecker.erl

```diff
@@ -4617,8 +4617,14 @@ check_guard_expression(Env, Guard) ->
 -spec check_guard(env(), list()) -> env().
 check_guard(Env, GuardSeq) ->
     RefTys = lists:foldl(fun(Guard, E) ->
-                             Ne = check_guard_expression(E, Guard),
-                             union_var_binds([Ne], E)
+                             NEnv = check_guard_expression(E, Guard),
+                             %% This is somewhat tricky.
+                             %% check_guard_expression/2 will return an env with just
+                             %% the types of vars used in the guard expression - we want these to
+                             %% override any types previously in the env.
+                             %% We still want to keep types of vars not used in the guard expression,
+                             %% though, so we pass both NEnv and E to union_var_binds/2.
+                             union_var_binds([NEnv, E], E)
```

Comment:
Does this override the type or use glb or lub if the same variable is in both envs?

Reply:
`union_var_binds` does GLB on the 1st argument, so the original type info from the spec is also taken into account.

Reply:
so maybe say "refine" instead of "override"?